File: studio/src/lib/stores/events.ts
import { listen } from '@tauri-apps/api/event';
import { setEntities, setSidecarStatus, setStaleAnchors } from './document.svelte';
import type { EntityDto, SidecarStatus, StaleAnchor } from '$lib/tauri/commands';

interface DocumentOpenedPayload {
  doc_id: string;
  entities: EntityDto[];
  sidecar_status: SidecarStatus;
}

interface EntitiesUpdatedPayload {
  doc_id: string;
  entities: EntityDto[];
}

interface SidecarStatusPayload {
  doc_id: string;
  status: SidecarStatus;
}

interface StaleAnchorsPayload {
  doc_id: string;
  anchors: StaleAnchor[];
}

interface ErrorPayload {
  doc_id: string;
  message: string;
}

let unlisteners: (() => void)[] = [];

export async function setupEventListeners() {
  unlisteners.push(
    await listen<DocumentOpenedPayload>('document-opened', (event) => {
      setEntities(event.payload.entities);
      setSidecarStatus(event.payload.sidecar_status);
    }),

    await listen<EntitiesUpdatedPayload>('entities-updated', (event) => {
      setEntities(event.payload.entities);
    }),

    await listen<SidecarStatusPayload>('sidecar-status', (event) => {
      setSidecarStatus(event.payload.status);
    }),

    await listen<StaleAnchorsPayload>('stale-anchors', (event) => {
      setStaleAnchors(event.payload.anchors);
    }),

    await listen<ErrorPayload>('parse-error', (event) => {
      console.warn('[Sparkdown] Parse error:', event.payload.message);
    }),

    await listen<ErrorPayload>('sidecar-error', (event) => {
      console.warn('[Sparkdown] Sidecar error:', event.payload.message);
    }),
  );
}

export function teardownEventListeners() {
  unlisteners.forEach((fn) => fn());
  unlisteners = [];
}

File: studio/src-tauri/src/events.rs
use tauri::{AppHandle, Emitter};

use crate::types::{DocId, EntityDto, SidecarStatus, StaleAnchor};

#[derive(Debug, Clone, serde::Serialize)]
pub struct DocumentOpenedPayload {
    pub doc_id: DocId,
    pub entities: Vec<EntityDto>,
    pub sidecar_status: SidecarStatus,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct EntitiesUpdatedPayload {
    pub doc_id: DocId,
    pub entities: Vec<EntityDto>,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct SidecarStatusPayload {
    pub doc_id: DocId,
    pub status: SidecarStatus,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct StaleAnchorsPayload {
    pub doc_id: DocId,
    pub anchors: Vec<StaleAnchor>,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct ErrorPayload {
    pub doc_id: DocId,
    pub message: String,
}

pub fn emit_document_opened(app: &AppHandle, payload: DocumentOpenedPayload) {
    let _ = app.emit("document-opened", payload);
}

pub fn emit_entities_updated(app: &AppHandle, payload: EntitiesUpdatedPayload) {
    let _ = app.emit("entities-updated", payload);
}

pub fn emit_sidecar_status(app: &AppHandle, payload: SidecarStatusPayload) {
    let _ = app.emit("sidecar-status", payload);
}

pub fn emit_stale_anchors(app: &AppHandle, payload: StaleAnchorsPayload) {
    let _ = app.emit("stale-anchors", payload);
}

pub fn emit_parse_error(app: &AppHandle, payload: ErrorPayload) {
    let _ = app.emit("parse-error", payload);
}

pub fn emit_sidecar_error(app: &AppHandle, payload: ErrorPayload) {
    let _ = app.emit("sidecar-error", payload);
}
